File: _Practice/LeetCode/018. Letter Combinations of a Phone Number (bfs).cpp
// bfs
// https://leetcode.com/problems/letter-combinations-of-a-phone-number/
// Time: O(4^n), where n is the number of digits

class Solution {
    const vector <string> pad = {
            "abc", "def", "ghi", "jkl",
            "mno", "pqrs", "tuv", "wxyz"
    };
public:
    vector <string> letterCombinations(string digits) {
        int n = digits.size();
        if (!n) return {};
        vector <string> res;

        queue <string> q;
        q.push("");

        for (char d: digits) {
            int SIZE = q.size();
            while (SIZE--) {
                string out = q.front();
                q.pop();
                for (char c: pad[d - '2']) {
                    q.push(out + c);
                }
            }
        }
        while (!q.empty()) {
            res.push_back(q.front());
            q.pop();
        }
        return res;
    }
};

File: _Practice/LeetCode/018. Letter Combinations of a Phone Number (bfs).java
// https://leetcode.com/problems/letter-combinations-of-a-phone-number/
// Time: O(4^n), where n is the number of digits

class Solution {
    private final String[] pad = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        if (digits.length() == 0) return res;
        Queue<String> q = new LinkedList<>();
        q.add("");
        int n = digits.length();
        int SIZE = 1, i = 0;
        while (q.peek().length() < n) {
            while (SIZE > 0) {
                String cur = q.remove();
                SIZE--;
                for (char c : pad[digits.charAt(i) - '2'].toCharArray()) {
                    q.add(cur + c);
                }
            }
            SIZE = q.size();
            i++;
        }
        res.addAll(q);
        return res;
    }
}
